cidc_api/utils/migrations.py

```diff
@@ -0,0 +1,159 @@
+    session = Session(bind=op.get_bind())
+
+    # Initialize an empty list of pieces of work to do on GCS resources
+    gcs_tasks = RollbackableQueue()
+
+    # Migrate all trial records
+    trials: List[TrialMetadata] = session.query(TrialMetadata).with_for_update().all()
+    for trial in trials:
+        migration = metadata_migration(trial.metadata_json)
+
+        # Update the trial metadata object
+        trial.metadata_json = migration.result
+
+        # Update the relevant downloadable files and GCS objects
+        for old_gcs_uri, artifact in migration.file_updates.items():
+            # Update the downloadable file associated with this blob
+            df: DownloadableFiles = (
+                session.query(DownloadableFiles)
+                .filter_by(object_url=old_gcs_uri)
+                .with_for_update()
+                .one()
+            )
+            for column, value in artifact.items():
+                if hasattr(df, column):
+                    setattr(df, column, value)
+
+            # Regenerate additional metadata from the migrated clinical trial
+            # metadata object.
+            df.additional_metadata = _get_uuid_info(
+                migration.result, artifact["upload_placeholder"]
+            )[1]
+
+            # If the GCS URI has changed, rename the blob
+            new_gcs_uri = artifact["object_url"]
+            if old_gcs_uri != new_gcs_uri:
+                renamer = PieceOfWork(
+                    lambda: rename_gcs_blob(
+                        GOOGLE_DATA_BUCKET, old_gcs_uri, new_gcs_uri
+                    ),
+                    lambda: rename_gcs_blob(
+                        GOOGLE_DATA_BUCKET, new_gcs_uri, old_gcs_uri
+                    ),
+                )
+                gcs_tasks.schedule(renamer)
+
+    # Migrate all assay upload successes
+    successful_assay_uploads: List[AssayUploads] = session.query(
+        AssayUploads
+    ).filter_by(status=AssayUploadStatus.MERGE_COMPLETED.value).with_for_update().all()
+    for upload in successful_assay_uploads:
+        migration = metadata_migration(upload.assay_patch)
+
+        # Update the metadata patch
+        upload.assay_patch = migration.result
+
+        # Update the GCS URIs of files that were part of this upload
+        old_file_map = upload.gcs_file_map
+        new_file_map = {}
+        for (
+            old_upload_uri,
+            old_target_uri,
+            artifact_uuid,
+        ) in upload.upload_uris_with_data_uris_with_uuids():
+            upload_timestamp = old_upload_uri[len(old_target_uri) + 1 :]
+            if old_target_uri in migration.file_updates:
+                new_target_uri = migration.file_updates[old_target_uri]["object_url"]
+                if old_gcs_uri != new_target_uri:
+                    new_upload_uri = "/".join([new_target_uri, upload_timestamp])
+                    renamer = PieceOfWork(
```

Comment:
wait, I though we decided not to move stuff in the upload bucked?

Reply:
I know we went back and forth on that. Here's my reasoning: migrating the `assay_uploads`  URIs and metadata patches without updating GCS would mean effectively losing the original files associated with a given upload from the system's perspective; since we decided to migrate `assay_uploads` metadata and URIs, I decided to also move stuff in the upload bucket.